6주차_ 과제1 [레스토랑 상세 페이지 구현하기]

seong-dodo opened the pull request:
안녕하세요. 

6주차 과제 : 레스토랑 상세 페이지 구현하기

- 라우팅의 이해
-  React router 사용하기

이번 과제에서는 `테스트 시나리오 네이밍`과 `커밋 메시지` 작성에 좀 더 신경 써서 작성해보도록 하겠습니다.
그리고 피드백 주시는 부분도 최대한 반영할 수 있도록 과제를 진행 하겠습니다.

> 1차 라우터 작업
- RestaurantsPage로 관심사 분리 
- MyComponent 만들기
- HomePage 만들기
- AboutPage 만들기
- NotFoundPage 만들기

> 2차 React Router 라이브러리 이용해서 고도화 작업
- 레스토랑 상세 페이지 구현하기


이렇게 차근차근 진행 해보도록 하겠습니다

seong-dodo committed:
webpack devserver 설정

- historyApiFallback 추가

seong-dodo committed:
RestaurantsPage로 관심사 분리

- RestaurantsPage로 관심사 분리 및 테스트 코드 분리
- App 컴포넌트와 테스트 코드 수정

seong-dodo committed:
HomePage에 대한 테스트 코드를 추가한다

- renders HomePage 테스트 추가

seong-dodo committed:
HomePage를 구현한다

- HomePage 화면 뼈대를 구현

seong-dodo committed:
AboutPage에 대한 테스트 코드를 추가한다

- renders AboutPage 테스트 추가

seong-dodo committed:
AboutPage를 구현한다

- AboutPage 화면 뼈대를 구현

seong-dodo committed:
NotFoundPage에 대한 테스트를 추가한다

- rendser NotFoundPage 테스트 추가

seong-dodo committed:
NotFoundPage를 구현한다

- NotFoundPage 화면 뼈대 구현

seong-dodo committed:
리액트 라우터 라이브러리 설치(고도화)하기 &  App에 대한 테스트를 추가한다

- 리액트 라우터 라이브러리 설치
- BroxserRouter, Switch 적용
- App 컴포넌트에서 HomePage 경로에 대한 화면 그리기 테스트 코드 추가

seong-dodo committed:
react router 라이브러리 접목하여 App컴포넌트 리팩토링을 한다

- HomePage에 대해 리액트 라우터 라이브러리 적용

seong-dodo committed:
HomePage에 대한 테스트 추가한다

- 리액트 라우터 라이브러리 추가에 대한 테스트 추가

seong-dodo committed:
AboutPage에 대한 테스트 추가한다

- 리액트 라우터 라이브러리 추가에 대한 테스트 추가

seong-dodo committed:
Homepage 컴포넌트에 react router 라이브러리 적용

- 리액트 라우터 라이브러리 이용해서 변경

seong-dodo committed:
App에 대한 테스트를 추가한다

- App 컴포넌트에서 AboutPage 경로에 대한 화면 그리기 테스트 코드 추가

seong-dodo committed:
react router 라이브러리 접목하여 App컴포넌트 리팩토링을 한다

- AboutPage에 대해 리액트 라우터 라이브러리 적용

seong-dodo committed:
RestaurantsPage에 대한 테스트 추가한다

- 리액트 라우터 라이브러리 추가에 대한 테스트 추가

seong-dodo committed:
App에 대한 테스트를 추가한다

- App 컴포넌트에서 RestaurantsPage 경로에 대한 화면 그리기 테스트 코드 추가

seong-dodo committed:
react router 라이브러리 접목하여 App컴포넌트 리팩토링을 한다

- RestaurantsPage에 대해 리액트 라우터 라이브러리 적용

seong-dodo committed:
NotFoundPage에 대한 테스트 추가한다

- 리액트 라우터 라이브러리 추가에 대한 테스트 추가

seong-dodo committed:
NotFoundPage 수정

seong-dodo committed:
App에 대한 테스트를 추가한다

- App 컴포넌트에서 NotFoundPage 경로에 대한 화면 그리기 테스트 코드 추가

seong-dodo committed:
react router 라이브러리 접목하여 App컴포넌트 리팩토링을 한다

- NotFoundPage에 대해 리액트 라우터 라이브러리 적용

seong-dodo committed:
App컴포넌트 header 추가하기

- 홈페이지로 연결되는 헤더 추가하기

Kihyun92 commented on src/App.test.jsx:
```diff
@@ -4,31 +4,78 @@ import { render } from '@testing-library/react';
+  });
 
-  expect(dispatch).toBeCalled();
+  context('without path', () => {
```
잘못된 경로로 접근한 경우를 테스트해주셨군요 👍🏼 

`without path`도 좋지만, `with invalid path`와 같이 적절하지 않다는 의미를 더 드러내줘도 좋을 것 같아요~

seong-dodo replied:
네 수정하겠습니다!! 코드를 작성하다보니 말씀해주신 `with invalid path`가 더 적절한 것 같아요~~ 👍

Kihyun92 commented on src/RestaurantsPage.jsx:
```diff
@@ -0,0 +1,27 @@
+export default function RestaurantsPage() {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(loadInitialData());
+  });
```
useEffect hook 에 두번째 인자를 넘겨주어서 보다 의도하는 경우에 실행해줄 수 있습니다.

초기 데이터는 최초 렌더시 한번만 호출해주기위해 두번째 인자로 빈 배열을 넘겨주는건 어떨까요?

- https://ko.reactjs.org/docs/hooks-reference.html#useeffect

seong-dodo replied:
넵 감사합니다~ 피드백 반영하여 수정하였습니다 😄 😄 😄

Kihyun92 reviewed (changes requested):
빠른 PR 너무 좋네요 👍🏼 👍🏼 

TDD cycle도 잘 지키며 해주시는 것 같습니다😄

seong-dodo committed:
RestaurantsContainer에 대한 테스트 코드를 추가한다

- 레스토랑목록 상세페이지에 대한 테스트 코드 추가

seong-dodo committed:
RestaurantsContainer에 레스토랑 상세 페이지 연결을 구현한다

- 레스토랑목록 상세페이지 Link를 연결한다

seong-dodo commented:
안녕하세요.

제가 `RestaurantsContainer`에 대한 테스트 코드 추가하고 기능을 구현해보았습니다.
 `RestaurantsContainer` 컴포넌트에서 레스토랑 목록에 <Link />를 연결해서 `RestaurantDetilPage`로 이동하는 부분에 대한 테스트 코드를 작성을 하였습니다. 제가 생각할 때 path : `/restaurants/Id`를 만들어줄거라고 정의는 했으나 유효하지 않은 경로니깐 `NotFoundPage`로 넘어갈 것 같아서  아래와 같이 `'Not Found'` 텍스트가 보일 것이라고 생각하고 테스트를 만들었습니다.

```javascript
context('with path /restaurants/Id', () => {
    it('renders restaurantDetilPage', () => {
      const { container } = render((
        <MemoryRouter initialEntries={['/restaurants/1']}>
          <RestaurantsContainer />
        </MemoryRouter>
      ));

      expect(container).toHaveTextContent('Not Found');
      // expect(container).toHaveTextContent('마법사주방');
      // expect(container).toHaveTextContent('메뉴');
    });
  });
```

그런데 아래와 같은 에러가 발생하였습니다.
제가 생각했던데로 `NotFoundPage`로 연결이 되지 않았다고 알려주는데 제가 테스트 코드를 잘 못 작성한 거 같습니다.
제가 깨우치지 못한 부분이 무엇인지 조언 부탁드립니다 🙏 
![image](https://user-images.githubusercontent.com/68280407/121031695-8c498a00-c7e5-11eb-9b3a-fda77f320df2.png)

seong-dodo committed:
App에 대한 테스트 코드를 추가한다

- App 컴포넌트에서 RestaurantsDetailPage 경로에 대한 화면 그리기 테스트 코드 추가

seong-dodo committed:
App 컴포넌트에 레스토랑 상세페이지를 라우트 한다 & 레스토랑컨테이너컴포넌트 수정

- RestaurantDetailPage의 경로를 라우트 해주기
- 레스토랑컨테이너컴포넌트 테스트 오류 코드 수정 (추후 재반영 할 예정)
- 시나리오 네이밍 등 오타 발견 및 수정

seong-dodo committed:
RestaurantDetailPage에 대한 테스트 코드를 추가한다

- renders RestaurantDetailPage 테스트 코드 추가

seong-dodo committed:
RestaurantDetailPage에 대한 뼈대 만들기 & 레스토랑 상세 정보보기 API 받아오기

- RestaurantDetailPage에 대한 뼈대 구현
- 레스토랑 상세 정보보기 API 받아오는 fetch함수 만들기

seong-dodo committed:
레스토랑 상세보기 API 관련 파일 추가

- api 테스트 코드 추가 (fetchRestaurantInformation)
- api 모킹 함수 추가 (fetchRestaurantInformation)
- fixtures 파일 생성 (fetchRestaurantInformation)

seong-dodo committed:
RestaurantDetailPage을 구현한다 & 오류 수정 등

- RestaurantDetailPage 기능을 구현
- 오류 수정

seong-dodo committed:
App 테스트 코드를 수정한다

- 테스트코드 작성 오류로 테스트 코드 수정

seong-dodo committed:
RestaurantDetailPage 테스트 코드를 수정한다

- 테스트코드 작성 오류로 테스트 코드 수정

Kihyun92 commented on src/RestaurantDetailPage.jsx:
```diff
@@ -0,0 +1,41 @@
+  });
+
+  useEffect(async () => {
+    const resultInformation = await fetchRestaurantInformation(props.match.params.id);
```
`useParams` hook 을 사용하면 조금 더 편하게 match.param 정보를 가져올 수 있습니다~

- https://reactrouter.com/web/api/Hooks/useparams

seong-dodo replied:
감사합니다. 수정해보았습니다 👍 😄

Kihyun92 commented on src/RestaurantDetailPage.jsx:
```diff
@@ -0,0 +1,41 @@
+} from './services/api';
+
+export default function RestaurantDetailPage(props) {
+  const [state, setState] = useState({
+    id: -1,
+    categoryId: -1,
+    name: '',
+    address: '',
+    menuItems: [],
+    reviews: [],
+    information: '',
+  });
```
아직 작업 중이신 것 같지만, 추후 스토어에서 관리해도 좋을 것 같네요~

seong-dodo replied:
넵! 스토어로 관리할 수 있도록 변경해보겠습니다 😆

Kihyun92 commented on src/RestaurantsContainer.jsx:
```diff
@@ -9,7 +13,9 @@ export default function RestaurantsContainer() {
     <ul>
       {restaurants.map((restaurant) => (
         <li key={restaurant.id}>
-          {restaurant.name}
+          <Link to={`/restaurants/${restaurant.id}`}>
+            {restaurant.name}
+          </Link>
         </li>
```
구조 분해 할당을 사용해도 좋을 것 같군요~

seong-dodo replied:
넵~ 감사합니다! 구조 분해 할당을 사용해보았습니다~~

Kihyun92 commented:
> 안녕하세요.
> 
> 제가 `RestaurantsContainer`에 대한 테스트 코드 추가하고 기능을 구현해보았습니다.
> `RestaurantsContainer` 컴포넌트에서 레스토랑 목록에 를 연결해서 `RestaurantDetilPage`로 이동하는 부분에 대한 테스트 코드를 작성을 하였습니다. 제가 생각할 때 path : `/restaurants/Id`를 만들어줄거라고 정의는 했으나 유효하지 않은 경로니깐 `NotFoundPage`로 넘어갈 것 같아서 아래와 같이 `'Not Found'` 텍스트가 보일 것이라고 생각하고 테스트를 만들었습니다.
> 
> ```js
> context('with path /restaurants/Id', () => {
>     it('renders restaurantDetilPage', () => {
>       const { container } = render((
>         <MemoryRouter initialEntries={['/restaurants/1']}>
>           <RestaurantsContainer />
>         </MemoryRouter>
>       ));
> 
>       expect(container).toHaveTextContent('Not Found');
>       // expect(container).toHaveTextContent('마법사주방');
>       // expect(container).toHaveTextContent('메뉴');
>     });
>   });
> ```
> 
> 그런데 아래와 같은 에러가 발생하였습니다.
> 제가 생각했던데로 `NotFoundPage`로 연결이 되지 않았다고 알려주는데 제가 테스트 코드를 잘 못 작성한 거 같습니다.
> 제가 깨우치지 못한 부분이 무엇인지 조언 부탁드립니다 🙏
> ![image](https://user-images.githubusercontent.com/68280407/121031695-8c498a00-c7e5-11eb-9b3a-fda77f320df2.png)

여러 테스트를 고려해보시는 것 같아 좋네요 👍🏼 👍🏼 

`App` 컴포넌트의 경우엔 유효하지 않은 경로로 접근시 라우트 처리가 되지만, `RestaurantsContainer`에서는 path 에 따른 라우팅 처리가 없기 때문에 그렇습니다~ 

단순히 초기 path 값을 통해서는 의도하신대로 처리할 수 없을 것 같아요.

seong-dodo committed:
구조분해할당 피드백 반영

seong-dodo committed:
reducer에 대한 테스트를 추가한다 & 레스토랑 상세 페이지 컴포넌트 변경

- setRestaurantDetail에 대한 테스트 추가
- RestaurantDetailPage 컴포넌트 변경

seong-dodo committed:
reducer, actions에 코드를 추가한다

- setRestaurantDetail에 대한 리듀서와 액션 코드 추가

seong-dodo committed:
네이밍 수정

- 레스토랑 상세페이지에 관련된 fetch함수와 load 액션명 통일(~Information -> ~Detail 로 변경)

seong-dodo committed:
actions에 대한 테스틑를 추가한다

- loadRestaurantDetail에 대한 테스트 추가

seong-dodo committed:
actions을 구현한다

- loadRestaurantDetail에 대한 actions을 구현한다

seong-dodo committed:
RestaurantDetailPage 컴포넌트 주석 처리 제거

seong-dodo committed:
피드백 반영 : 구조분해 할당

- 피드백내용 재검토하여 구조분해 할당 필요한 부분 수정
- CategoriesContainer, RegionsContainer 컴포넌트에 구조분해 할당 적용
- RestaurantDetailPage 컴포넌트 적용x

seong-dodo committed:
App에 대한 테스트를 추가한다

- 리듀서에서 관리하는 상태(restaurantDetail)에 대한 모킹함수 추가

seong-dodo committed:
App에 대한 테스트를 추가한다

- renders RestaurantDetailPage 테스트 내용 추가

seong-dodo committed:
RestaurantDetailPage에 대한 테스트 코드를 수정한다

- without restaurantDetail일 경우에 대한 테스트 코드를 추가

seong-dodo committed:
RestaurantDetailPage에 대한 테스트 코드를 수정한다

- with restaurantDetail일 경우에 대한 테스트 코드를 추가
- 테스트커버리지 100%
- ESLint 통과
- 모든 인수 테스트 통과

Kihyun92 commented on src/App.test.jsx:
```diff
@@ -4,31 +4,95 @@ import { render } from '@testing-library/react';
+  MemoryRouter,
+} from 'react-router-dom';
+
+import RESTAURANDETAIL from '../fixtures/restaurantDetail';
```
대문자로 표현 할거라면 `RESTAURANT_DETAIL`과 같이 구분해주는건 어떨까요? 😄

seong-dodo replied:
수정해보았습니다. 오타도 있었네요 ㅜㅜ 감사합니다 👍 😄👍 😄👍 😄

Kihyun92 commented on src/RestaurantDetailPage.jsx:
```diff
@@ -0,0 +1,46 @@
+
+  const restaurantDetail = useSelector(get('restaurantDetail'));
+
+  return (restaurantDetail === null
+    ? <div>404 Not Found</div>
+    : (
+      <div>
+        <h1>{restaurantDetail.name}</h1>
+        <p>
+          주소:
+          {' '}
+          {restaurantDetail.address}
+        </p>
+        <h2>메뉴</h2>
+        <ul>
+          {restaurantDetail.menuItems.map((menuItem) => (
+            <li key={menuItem.id}>
+              {menuItem.name}
+            </li>
+          ))}
+        </ul>
+      </div>
+    )
+  );
```
삼항연산자로 레스토랑 정보가 없는 경우에 대해서 처리해주신 것 같아요!

```javascript
if (restaurantDetail === null) {
  return <div>404 Not Found</div>
}

return (
   <div>
        <h1>{restaurantDetail.name}</h1>
        <p>
          주소:
          {' '}
          {restaurantDetail.address}
        </p>
        <h2>메뉴</h2>
        <ul>
          {restaurantDetail.menuItems.map((menuItem) => (
            <li key={menuItem.id}>
              {menuItem.name}
            </li>
          ))}
        </ul>
      </div>
    );
```
위 예시와 같이 `Guard Clauses`를 사용해서 처리해주는 것도 좋을 것 같네요~

- https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

seong-dodo replied:
안녕하세요. 
피드백을 받고 삼항연산자로 표현한 부분을 바꾸어 보았습니다.
그런데 말씀해주신 예시처럼 함수를 만들어서 변경하지 않고😅😅😅, 기존에 만들었던 `NotFoundPage`를 활용해보았습니다;;;

생각을 해보니 Page라는 이름이 붙은 컴포넌트들은 각각 프레젠테이션컴포넌트로써 역활을 하고 있는데,  현재 제가 만든`restaurantDetailPage`에는 로직이 포함되어 있어 컨테이너 컴포넌트 역활을 하고있습니다.
컨테이너 컴포넌트로 분리하는게 좋을지.. 아니면 말씀해주신 `Guard Clauses`을 사용해서 `restaurantDetailPage` 컴포넌트에서 모든걸 처리하도록 표현해야할지 고민이 되어 글 남겨봅니다.

```
function getPayAmount() {
  if (isDead) return deadAmount();
  if (isSeparated) return separatedAmount();
  if (isRetired) return retiredAmount();
  return normalPayAmount();
}
```

Kihyun92 replied:
저 조건에 걸리는 경우가 not found 도 있지만, 데이터를 로딩해오는 경우도 있을 것 같아요~ 이 경우엔 NotFound 페이지를 보여주는게 맞을지 고민해봐도 좋을 것 같네요 😊

seong-dodo replied:
아항!! 감사합니다 😀 😀 😀 
그럼 로딩해주는 경우로 생각을 해보겠습니다 👍 👍  👍  👍  👍

Kihyun92 commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -2,18 +2,38 @@ import { render } from '@testing-library/react';
+  });
 
-  expect(container).toHaveTextContent('마법사주방');
+  context('with path /restaurants/:id', () => {
```
context는 with path <=> without path와 같이 문맥이 서로 다른 경우에 사용합니다.

그외의 경우에는 describe로 처리해도 좋을 것 같아요~

seong-dodo replied:
넵 ~ 감사합니다 😃

Kihyun92 reviewed (changes requested):
잘 진행하고 계신 것 같군요! 👍🏼👍🏼

seong-dodo committed:
피드백 반영

- 대문자 표현 수정

seong-dodo committed:
피드백 반영

- 불필요한 context을 describe로 수정

seong-dodo committed:
피드백 반영

- 삼항연산자 제거

Kihyun92 commented on src/App.jsx:
```diff
@@ -1,32 +1,32 @@
+        </h1>
+      </header>
+      <Switch>
+        <Route exact path="/" component={HomePage} />
+        <Route path="/about" component={AboutPage} />
+        <Route exact path="/restaurants" component={RestaurantsPage} />
+        <Route path="/restaurants/:id" component={RestaurantDetailPage} />
+        <Route component={NotFoundPage} />
```
전반적으로 `**Page`라는 처리가 필요할까요? (모두 다 어색하다는 말은 아닙니다 😅)

page 들이라는 의도를 드러내고 싶다면 pages 폴더와 같이 처리해주고 컴포넌트 네이밍은 역할만 드러내 주는 방법도 있을 것 같아요~

seong-dodo replied:
아항😀 감사합니다. 그런방법이 있었군요!!!👍 👍 👍 👍 👍

Kihyun92 commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,60 @@
+    });
+
+    it('renders RestaurantsPage and Not Found', () => {
+      const { queryByText } = render((
+        <MemoryRouter initialEntries={['/restaurants/1']}>
+          <RestaurantDetailPage />
+        </MemoryRouter>
+      ));
```
반복되는 렌더 관련 로직은 함수로 추출해줘도 좋을 것 같아요~

seong-dodo replied:
넵 😀  수정해보겠습니다.

seong-dodo committed:
피드백 반영

- 삼항연산자 대신 Guard Clauses 적용
- not found 경우를 데이터 로딩해오는 경우로 변경

seong-dodo committed:
RestaurantDetailPage에 대한 테스트를 수정한다

- renders loading state에 대한 테스트로 수정

seong-dodo committed:
피드백 반영

- 반복되는 렌더 로직 함수로 수정

Kihyun92 reviewed (approved):
이번주도 고생하셨습니다~
TDD에 많이 익숙해지신 것 같아요 ㅎㅎ

과제 풀이 영상 꼭 봐보시고, 이번주에 어떤걸 배웠는지, 어떤걸 더 배우고 싶은지 정리해보셔도 좋을 것 같네요 😊

슬랙 채널에 회고 올려주시면 머지하겠습니다~~
